WMS error with CRS:84

Making a MrGeo WMS request like causes an exception (pasted below)

/mrgeo/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&CRS=CRS:84&BBOX=77.946517882089964,27.716680637049464,87.00737985470802,33.50020955574184&WIDTH=1128&HEIGHT=720&LAYERS=my-image-name&STYLES=&FORMAT=image/png&BGCOLOR=0xFEFFFF&TRANSPARENT=TRUE

Exception:

```
MrGeo Web Map Service:   WMS service exceptions:Service Error Description: org.mrgeo.services.mrspyramid.rendering.ImageRendererException: java.lang.RuntimeException: OGR Error: Unsupported SRS
                at org.mrgeo.services.mrspyramid.rendering.ImageRendererAbstract.renderImage(ImageRendererAbstract.java:422)
                at org.mrgeo.resources.wms.WmsGenerator.getMap(WmsGenerator.java:447)
                at org.mrgeo.resources.wms.WmsGenerator.handleRequest(WmsGenerator.java:300)
                at org.mrgeo.resources.wms.WmsGenerator.doGet(WmsGenerator.java:160)
                at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.lang.reflect.Method.invoke(Method.java:498)
                at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
                at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
                at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
                at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
                at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
                at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
                at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
                at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
                at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
                at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
                at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
                at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
                at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
                at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
                at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
                at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
                at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
                at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
                at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
                at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
                at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
                at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
                at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
                at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
                at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
                at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
                at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
                at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
                at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
                at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
                at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
                at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
                at org.eclipse.jetty.server.Server.handle(Server.java:499)
                at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
                at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
                at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
                at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
                at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
                at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: OGR Error: Unsupported SRS
                at org.gdal.osr.osrJNI.SpatialReference_ImportFromEPSG(Native Method)
                at org.gdal.osr.SpatialReference.ImportFromEPSG(SpatialReference.java:549)
                at org.mrgeo.services.utils.RequestUtils.reprojectBoundsToWGS84(RequestUtils.java:114)
                at org.mrgeo.services.mrspyramid.rendering.ImageRendererAbstract.renderImage(ImageRendererAbstract.java:286)
                ... 46 more
```

--- a/mrgeo-core/src/main/java/org/mrgeo/geometry/Reprojector.java
+++ b/mrgeo-core/src/main/java/org/mrgeo/geometry/Reprojector.java
@@ -77,22 +77,22 @@ public static Reprojector createFromWkt(String wktSrc, String wktDest)
   return new Reprojector(sourceSrs, destSrs);
 }
 
-private static int parseEpsgCode(String epsg)
-{
-  String prefix = "epsg:";
-  int index = epsg.toLowerCase().indexOf(prefix);
-  if (index >= 0)
-  {
-    try
-    {
-      return Integer.parseInt(epsg.substring(index + prefix.length()));
-    }
-    catch (NumberFormatException nfe)
-    {
-    }
-  }
-  throw new IllegalArgumentException("Invalid EPSG code: " + epsg);
-}
+//private static int parseEpsgCode(String epsg)
+//{
+//  String prefix = "epsg:";
+//  int index = epsg.toLowerCase().indexOf(prefix);
+//  if (index >= 0)
+//  {
+//    try
+//    {
+//      return Integer.parseInt(epsg.substring(index + prefix.length()));
+//    }
+//    catch (NumberFormatException nfe)
+//    {
+//    }
+//  }
+//  throw new IllegalArgumentException("Invalid EPSG code: " + epsg);
+//}
 
 @Override
 public void filter(WritablePoint p)

--- a/mrgeo-core/src/main/java/org/mrgeo/geometry/Reprojector.java
+++ b/mrgeo-core/src/main/java/org/mrgeo/geometry/Reprojector.java
@@ -50,11 +50,9 @@ private Reprojector(SpatialReference sourceSrs, SpatialReference destSrs)
 public static Reprojector createFromCode(String codeSrc, String codeDest)
 {
   SpatialReference sourceSrs = new SpatialReference();
-  int sourceEpsg = parseEpsgCode(codeSrc);
-  sourceSrs.ImportFromEPSG(sourceEpsg);
+  sourceSrs.SetFromUserInput(codeSrc);
   SpatialReference destSrs = new SpatialReference();
-  int destEpsg = parseEpsgCode(codeDest);
-  destSrs.ImportFromEPSG(destEpsg);
+  destSrs.SetFromUserInput(codeDest);
 
   return new Reprojector(sourceSrs, destSrs);
 }

--- a/mrgeo-services/mrgeo-services-core/src/main/java/org/mrgeo/services/mrspyramid/rendering/ImageRendererAbstract.java
+++ b/mrgeo-services/mrgeo-services-core/src/main/java/org/mrgeo/services/mrspyramid/rendering/ImageRendererAbstract.java
@@ -161,22 +161,22 @@ private static MrsImage getImageForScale(final MrsPyramid pyramid, final double 
   return image;
 }
 
-private static int parseEpsgCode(String epsg)
-{
-  String prefix = "epsg:";
-  int index = epsg.toLowerCase().indexOf(prefix);
-  if (index >= 0)
-  {
-    try
-    {
-      return Integer.parseInt(epsg.substring(index + prefix.length()));
-    }
-    catch (NumberFormatException ignored)
-    {
-    }
-  }
-  throw new IllegalArgumentException("Invalid EPSG code: " + epsg);
-}
+//private static int parseEpsgCode(String epsg)
+//{
+//  String prefix = "epsg:";
+//  int index = epsg.toLowerCase().indexOf(prefix);
+//  if (index >= 0)
+//  {
+//    try
+//    {
+//      return Integer.parseInt(epsg.substring(index + prefix.length()));
+//    }
+//    catch (NumberFormatException ignored)
+//    {
+//    }
+//  }
+//  throw new IllegalArgumentException("Invalid EPSG code: " + epsg);
+//}
 
 /*
  * (non-Javadoc)

--- a/mrgeo-services/mrgeo-services-core/src/main/java/org/mrgeo/services/mrspyramid/rendering/ImageRendererAbstract.java
+++ b/mrgeo-services/mrgeo-services-core/src/main/java/org/mrgeo/services/mrspyramid/rendering/ImageRendererAbstract.java
@@ -360,7 +360,7 @@ public MrGeoRaster renderImage(final String pyramidName, final Bounds requestBou
           scrs.ImportFromEPSG(4326);
 
           SpatialReference dcrs = new SpatialReference();
-          dcrs.ImportFromEPSG(parseEpsgCode(epsg));
+          dcrs.SetFromUserInput(epsg);
 
           dstcrs = dcrs.ExportToWkt();
 

--- a/mrgeo-services/mrgeo-services-core/src/main/java/org/mrgeo/services/utils/RequestUtils.java
+++ b/mrgeo-services/mrgeo-services-core/src/main/java/org/mrgeo/services/utils/RequestUtils.java
@@ -59,18 +59,18 @@ public static Bounds boundsFromParam(final String param)
  * Reprojects a bounds to Geographic
  *
  * @param bounds the projected input bounds
- * @param epsg   the epsg string of the projected bounds crs
+ * @param srs   the spatial reference of the projected bounds crs. The format is described
+ *              in http://gdal.org/java/org/gdal/osr/SpatialReference.html#SetFromUserInput(java.lang.String).
+ *              Examples include "EPSG:4326" and "CRS:84".
  * @return geographic bounds
  */
-public static Bounds reprojectBounds(final Bounds bounds, final String epsg)
+public static Bounds reprojectBounds(final Bounds bounds, final String srs)
 {
-  if (epsg != null && !(epsg.equalsIgnoreCase("EPSG:4326")))
+  if (srs != null && !(srs.equalsIgnoreCase("EPSG:4326")))
   {
-
     SpatialReference src = new SpatialReference(GDALUtils.EPSG4326());
     SpatialReference dst = new SpatialReference();
-    String[] code = epsg.split(":");
-    dst.ImportFromEPSG(Integer.parseInt(code[1]));
+    dst.SetFromUserInput(srs);
 
     CoordinateTransformation tx = new CoordinateTransformation(src, dst);
 
@@ -99,38 +99,35 @@ public static Bounds reprojectBounds(final Bounds bounds, final String epsg)
  * Reprojects a bounds to WGS-84
  *
  * @param bounds the projected input bounds
- * @param epsg   the epsg string of the projected bounds crs
+ * @param srs   the spatial reference of the projected bounds crs. The format is
+ *              described in http://gdal.org/java/org/gdal/osr/SpatialReference.html#SetFromUserInput(java.lang.String)
+ *              Examples include "EPSG:4326" and "CRS:84".
  * @return geographic bounds
  */
-public static Bounds reprojectBoundsToWGS84(final Bounds bounds, final String epsg)
+public static Bounds reprojectBoundsToWGS84(final Bounds bounds, final String srs)
 {
-  if (epsg != null)
+  if (srs != null)
   {
-    String[] code = epsg.split(":");
-    int srcEpsg = Integer.parseInt(code[1]);
-    if (srcEpsg != 4326)
-    {
-      SpatialReference src = new SpatialReference();
-      src.ImportFromEPSG(srcEpsg);
+    SpatialReference src = new SpatialReference();
+    src.SetFromUserInput(srs);
 
-      SpatialReference dst = new SpatialReference(GDALUtils.EPSG4326());
-      CoordinateTransformation tx = new CoordinateTransformation(src, dst);
+    SpatialReference dst = new SpatialReference(GDALUtils.EPSG4326());
+    CoordinateTransformation tx = new CoordinateTransformation(src, dst);
 
-      double[] c1;
-      double[] c2;
-      double[] c3;
-      double[] c4;
+    double[] c1;
+    double[] c2;
+    double[] c3;
+    double[] c4;
 
-      c1 = tx.TransformPoint(bounds.w, bounds.s);
-      c2 = tx.TransformPoint(bounds.w, bounds.n);
-      c3 = tx.TransformPoint(bounds.e, bounds.s);
-      c4 = tx.TransformPoint(bounds.e, bounds.n);
+    c1 = tx.TransformPoint(bounds.w, bounds.s);
+    c2 = tx.TransformPoint(bounds.w, bounds.n);
+    c3 = tx.TransformPoint(bounds.e, bounds.s);
+    c4 = tx.TransformPoint(bounds.e, bounds.n);
 
-      return new Bounds(Math.min(Math.min(c1[0], c2[0]), Math.min(c3[0], c4[0])),
-          Math.min(Math.min(c1[1], c2[1]), Math.min(c3[1], c4[1])),
-          Math.max(Math.max(c1[0], c2[0]), Math.max(c3[0], c4[0])),
-          Math.max(Math.max(c1[1], c2[1]), Math.max(c3[1], c4[1])));
-    }
+    return new Bounds(Math.min(Math.min(c1[0], c2[0]), Math.min(c3[0], c4[0])),
+        Math.min(Math.min(c1[1], c2[1]), Math.min(c3[1], c4[1])),
+        Math.max(Math.max(c1[0], c2[0]), Math.max(c3[0], c4[0])),
+        Math.max(Math.max(c1[1], c2[1]), Math.max(c3[1], c4[1])));
   }
   return bounds.clone();
 }

--- a/mrgeo-services/mrgeo-services-wms/testFiles/org.mrgeo.resources.wms/BadRequestTest/wms-invalid-coordsys.xml
+++ b/mrgeo-services/mrgeo-services-wms/testFiles/org.mrgeo.resources.wms/BadRequestTest/wms-invalid-coordsys.xml
@@ -1,10 +1,10 @@
 <ServiceExceptionReport version="1.3.0">
-  <ServiceException><![CDATA[org.mrgeo.services.mrspyramid.rendering.ImageRendererException: java.lang.RuntimeException: OGR Error: Unsupported SRS
+  <ServiceException><![CDATA[org.mrgeo.services.mrspyramid.rendering.ImageRendererException: java.lang.RuntimeException: OGR Error: General Error
 	at org.mrgeo.services.mrspyramid.rendering.ImageRendererAbstract.renderImage(ImageRendererAbstract.java:422)
 	at org.mrgeo.resources.wms.WmsGenerator.getMap(WmsGenerator.java:447)
 	at org.mrgeo.resources.wms.WmsGenerator.handleRequest(WmsGenerator.java:300)
 	at org.mrgeo.resources.wms.WmsGenerator.doGet(WmsGenerator.java:160)
-	at sun.reflect.GeneratedMethodAccessor18.invoke(Unknown Source)
+	at sun.reflect.GeneratedMethodAccessor19.invoke(Unknown Source)
 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 	at java.lang.reflect.Method.invoke(Method.java:498)
 	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
@@ -29,10 +29,10 @@
 	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:591)
 	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:571)
 	at java.lang.Thread.run(Thread.java:745)
-Caused by: java.lang.RuntimeException: OGR Error: Unsupported SRS
-	at org.gdal.osr.osrJNI.SpatialReference_ImportFromEPSG(Native Method)
-	at org.gdal.osr.SpatialReference.ImportFromEPSG(SpatialReference.java:549)
-	at org.mrgeo.services.utils.RequestUtils.reprojectBoundsToWGS84(RequestUtils.java:114)
+Caused by: java.lang.RuntimeException: OGR Error: General Error
+	at org.gdal.osr.osrJNI.SpatialReference_SetFromUserInput(Native Method)
+	at org.gdal.osr.SpatialReference.SetFromUserInput(SpatialReference.java:457)
+	at org.mrgeo.services.utils.RequestUtils.reprojectBoundsToWGS84(RequestUtils.java:112)
 	at org.mrgeo.services.mrspyramid.rendering.ImageRendererAbstract.renderImage(ImageRendererAbstract.java:286)
 	... 28 more
 ]]></ServiceException>

--- a/mrgeo-services/mrgeo-services-wms/testFiles/org.mrgeo.resources.wms/GetMapTest/testGetMapInvalidLayer.xml
+++ b/mrgeo-services/mrgeo-services-wms/testFiles/org.mrgeo.resources.wms/GetMapTest/testGetMapInvalidLayer.xml
@@ -4,7 +4,7 @@
 	at org.mrgeo.resources.wms.WmsGenerator.getMap(WmsGenerator.java:447)
 	at org.mrgeo.resources.wms.WmsGenerator.handleRequest(WmsGenerator.java:300)
 	at org.mrgeo.resources.wms.WmsGenerator.doGet(WmsGenerator.java:160)
-	at sun.reflect.GeneratedMethodAccessor18.invoke(Unknown Source)
+	at sun.reflect.GeneratedMethodAccessor19.invoke(Unknown Source)
 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 	at java.lang.reflect.Method.invoke(Method.java:498)
 	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)

